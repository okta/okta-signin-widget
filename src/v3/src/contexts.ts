/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { createContext } from 'preact';
import { useContext } from 'preact/hooks';

import {
  ILayoutContext, IStepperContext, IWidgetContext, UISchemaLayoutType,
} from './types';

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<T | undefined>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };

export const StepperContext = createContext<IStepperContext>({
  stepIndex: 0,
  setStepIndex: () => {},
});

export const useStepperContext = () => useContext(StepperContext);

export const LayoutContext = createContext<ILayoutContext>({
  layoutDirection: UISchemaLayoutType.VERTICAL,
});

export const useLayoutContext = () => useContext(LayoutContext);
