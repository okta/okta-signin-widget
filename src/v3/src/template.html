<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Okta Sign-in Widget v3 Playground</title>
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <link rel="apple-touch-icon" href="/static/icons/apple-touch-icon.png">
    <script type="text/javascript" nonce="playground">
      // this script block can be removed after OKTA-517096
      window.cspNonce = 'playground';
    </script>
    <% preact.headEnd %>
  </head>
  <body>
    <div id="okta-login-container"></div>
    <% preact.bodyEnd %>
  </body>
</html>
