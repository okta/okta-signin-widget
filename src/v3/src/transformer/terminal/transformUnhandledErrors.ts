/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { AuthApiError, OAuthError } from '@okta/okta-auth-js';

import { getMessage } from '../../../../v2/ion/i18nTransformer';
import {
  FormBag,
  InfoboxElement,
  WidgetProps,
} from '../../types';
import { loc } from '../../util';
import { createForm } from '../utils';

type ErrorTransformer = (widgetProps: WidgetProps, error?: (AuthApiError | OAuthError)) => FormBag;
type ErrorTester<T extends (AuthApiError | OAuthError)> = {
  tester: (err?: T) => boolean,
  message: (err?: T) => string,
};

const getErrorMessage = (
  error?: (AuthApiError | OAuthError),
  widgetProps?: WidgetProps,
) : string => {
  const authApiErrorChecks: ErrorTester<AuthApiError>[] = [
    // error message comes from server response
    {
      tester: (err?: AuthApiError) => !!(err && err.xhr && !err.errorSummary && err.xhr.responseText?.includes('messages')),
      message: (err?: AuthApiError) => {
        const errorResponse = JSON.parse(err!.xhr!.responseText);
        const { messages: { value: [message] } } = errorResponse;

        // TODO: re-visit, handle side effects in hooks
        // If the session expired, this clears session to allow new transaction bootstrap
        if (widgetProps && message.i18n.key === 'idx.session.expired') {
          const { authClient } = widgetProps;
          authClient?.transactionManager.clear();
        }

        return getMessage(message);
      },
    },
    // special error messages
    {
      tester: (err?: AuthApiError) => !!err?.errorCode && !!err?.errorSummary,
      message: (err?: AuthApiError) => err!.errorSummary,
    },
  ];
  const oauthErrorChecks: ErrorTester<OAuthError>[] = [
    {
      tester: (err?: OAuthError) => err?.error === 'invalid_request' && err?.error_description === 'The recovery token is invalid',
      message: () => loc('oie.invalid.recovery.token', 'login'),
    },
    {
      tester: (err?: OAuthError) => err?.error === 'access_denied' && !!err?.error_description,
      message: () => loc('oie.feature.disabled', 'login'),
    },
    {
      tester: (err?: OAuthError) => !!err?.error && !!err?.error_description,
      message: () => loc('oie.configuration.error', 'login'),
    },
  ];

  // find the message that meets the tester condition
  let message: string | undefined;
  switch (error?.name) {
    case 'AuthApiError':
      message = authApiErrorChecks
        .find(({ tester }) => tester(error as AuthApiError))?.message(error as AuthApiError);
      break;
    case 'OAuthError':
      message = oauthErrorChecks
        .find(({ tester }) => tester(error as OAuthError))?.message(error as OAuthError);
      break;
    default:
      // intentionally fall through
  }
  // default fall back for unknown errors
  return message || loc('oform.error.unexpected');
};

export const transformUnhandledErrors: ErrorTransformer = (widgetProps, error) => {
  const formBag: FormBag = createForm();

  formBag.uischema.elements = [{
    type: 'InfoBox',
    options: {
      message: getErrorMessage(error, widgetProps),
      class: 'ERROR',
      dataSe: 'callout',
    },
  } as InfoboxElement];

  return formBag;
};
