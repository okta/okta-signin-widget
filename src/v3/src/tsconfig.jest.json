{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "sourceMap": true,
    "inlineSourceMap": false,
    "target": "ES2019"
  }
}
