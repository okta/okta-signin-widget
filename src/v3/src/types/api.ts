/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { MinimalOktaAuthIdxInterface, Tokens } from '@okta/okta-auth-js';

import {
  HooksAPI,
  RenderErrorCallback,
  RenderResult,
  RenderSuccessCallback,
  RouterEventsAPI,
} from '../../../types';
import { RenderOptions, WidgetOptions } from './widget';

// Keep a duplicate OktaSignInAPI interface because the version of auth-js differs between
// Gen3 and Gen2 so there are type mismatches in their exposed type interfaces
export interface OktaSignInAPI extends HooksAPI, RouterEventsAPI {
  // Gen3 only
  readonly options: Pick<WidgetOptions, 'brandName'>;

  // Gen3 supports only IDX
  authClient: MinimalOktaAuthIdxInterface;
  show(): void;
  hide(): void;
  remove(): void;
  showSignIn(options: RenderOptions): Promise<RenderResult>;
  showSignInToGetTokens(options: RenderOptions): Promise<Tokens>;
  showSignInAndRedirect(options: RenderOptions): Promise<void>;
  renderEl(
    options: RenderOptions,
    success?: RenderSuccessCallback,
    error?: RenderErrorCallback
  ): Promise<RenderResult>;

  getUser(): void
}
