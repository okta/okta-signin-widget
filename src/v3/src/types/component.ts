/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { FunctionComponent } from 'preact';
import { StateUpdater } from 'preact/hooks';

import { BlurHandler } from './handlers';
import {
  ButtonElement,
  FieldElement,
  UISchemaElement,
  WidgetMessage,
} from './schema';

export type TesterFunction = (
  uiElement: UISchemaElement | FieldElement | ButtonElement
) => boolean;

export type UISchemaElementComponentWithValidationProps = {
  uischema: FieldElement;
  type?: string;
  setTouched?: StateUpdater<boolean>,
  errors?: WidgetMessage[],
  setErrors?: StateUpdater<WidgetMessage[] | undefined>,
  onValidateHandler?: (
    setErrors?: StateUpdater<WidgetMessage[] | undefined>,
    value?: string | boolean | number,
  ) => void,
  handleChange?: (value: string | number | boolean) => void;
  handleBlur?: BlurHandler;
  describedByIds?: string;
};

export type UISchemaElementComponentProps = {
  uischema: UISchemaElement;
};

export type UISchemaElementComponent<
  T = Record<string, unknown>,
> = FunctionComponent<UISchemaElementComponentProps & T>;

export type Renderer = {
  tester: TesterFunction;
  renderer: UISchemaElementComponent;
};
