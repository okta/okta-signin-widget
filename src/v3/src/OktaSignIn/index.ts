/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { OktaAuth, OktaAuthOptions } from '@okta/okta-auth-js';
import pick from 'lodash/pick';
import { h, render } from 'preact';

import { Widget } from '../components/Widget';
import { JsonObject } from '../types';
import {
  OktaWidgetEventHandler,
  OktaWidgetEventType,
  RenderOptions,
  WidgetProps,
} from '../types/widget';

// TODO: Once SIW is merged into okta-signin-widget repo, remove these.
export type RenderSuccessCallback = {
  (res: JsonObject): void;
};
export type RenderErrorCallback = {
  (args: Error): void;
};
// TODO: Once SIW is merged into okta-signin-widget repo, remove these
export type RenderResult = JsonObject;
export type Tokens = JsonObject;

console.debug(`${OKTA_SIW_VERSION}-g${OKTA_SIW_COMMIT_HASH.substring(0, 7)}`);

export default class OktaSignIn {
  /**
   * Version string
   */
  readonly version: string;

  /**
   * Package version
   */
  static readonly __version: string = OKTA_SIW_VERSION;

  /**
   * Commit SHA
   */
  static readonly __commit: string = OKTA_SIW_COMMIT_HASH;

  /**
   * Okta Signin Widget options
   */
  readonly options: WidgetProps;

  /**
   * Instance of OktaAuth client
   */
  readonly authClient: OktaAuth;

  /**
   * Registered event handlers
   */
  readonly events: {
    [key in OktaWidgetEventType]: OktaWidgetEventHandler
  } | Record<string, never>;

  el: string | null;

  constructor(options: WidgetProps) {
    this.version = OKTA_SIW_VERSION;
    this.options = options;
    this.el = null;

    this.events = {};

    // if authClient is set, authParams are disregarded
    if (options.authClient) {
      // safety check
      if (!(options.authClient instanceof OktaAuth)) {
        throw new Error('expected "authClient" to be an instance of OktaAuth');
      }

      // instance of OktaAuth client
      this.authClient = options.authClient;
      // HACK: options should not be touched once instance is initialized
      // Update auth-js to preserve useGenericRemediator when passed from function level
      this.authClient.options.idx = this.authClient.options.idx || {};
      this.authClient.options.idx.useGenericRemediator = true;
    } else {
      const authParams = {
        ...pick(options, [
          'issuer',
          'clientId',
          'redirectUri',
          'state',
          'scopes',
          'flow',
          'codeChallenge',
          'codeChallengeMethod',
          'recoveryToken',
        ]),
        ...options.authParams,
        idx: {
          useGenericRemediator: true,
        },
      };

      // set default issuer using baseUrl
      if (!authParams.issuer) {
        authParams.issuer = `${options.baseUrl}/oauth2/default`;
      }

      // instance of OktaAuth client
      this.authClient = new OktaAuth(authParams);

      // add widget version to extended user agent header
      // eslint-disable-next-line no-underscore-dangle
      const userAgent = this.authClient._oktaUserAgent;
      if (userAgent) {
        userAgent.addEnvironment('okta-signin-widget-next');
        userAgent.addEnvironment(OKTA_SIW_COMMIT_HASH);
      }

      if (options.el) {
        this.renderEl({
          el: options.el,
          clientId: authParams.clientId,
          redirectUri: authParams.redirectUri,
        });
      }
    }
  }

  renderEl(
    options: RenderOptions,
    onSuccess?: RenderSuccessCallback,
    onError?: RenderErrorCallback,
  ): Promise<RenderResult> {
    const { el } = options;
    this.el = el;

    return new Promise<RenderResult>((resolve, reject) => {
      const onSuccessWrapper = (data: JsonObject): void => {
        onSuccess?.(data);
        resolve(data);
      };
      const onErrorWrapper = (err: Error): void => {
        onError?.(err);
        reject(err);
      };
      try {
        const target = typeof this.el === 'string'
          ? document.querySelector(this.el)
          : this.el;

        if (target) {
          // @ts-ignore OKTA-508744
          render(h(Widget, {
            // @ts-expect-error
            events: this.events,
            authClient: this.authClient,
            globalSuccessFn: onSuccessWrapper,
            globalErrorFn: onErrorWrapper,
            ...this.options,
          }), target);
        } else {
          throw new Error(`could not find element ${el}`);
        }
      } catch (error) {
        if (typeof onError === 'function') {
          onErrorWrapper(error as Error);
        }
        reject(error);
      }
    });
  }

  showSignInToGetTokens(options = {}): Promise<Tokens> {
    if (this.authClient.isAuthorizationCodeFlow() && this.authClient.isPKCE()) {
      throw new Error('"showSignInToGetTokens()" should not be used for authorization_code flow. Use "showSignInAndRedirect()" instead');
    }

    return this.renderEl({
      ...this.buildRenderOptions(options),
      redirect: 'never',
    })
      .then((res) => res.tokens as JsonObject); // Remove this cast once merged into okta-signin-widget
  }

  showSignInAndRedirect(options = {}): Promise<void> {
    // This method should never return, it will either redirect or reject with error
    return this.renderEl({
      ...this.buildRenderOptions(options),
      redirect: 'always',
    })
      .then(() => { });
  }

  showSignIn(options = {}): Promise<RenderResult> {
    return this.renderEl(
      this.buildRenderOptions(options),
    );
  }

  // eslint-disable-next-line class-methods-use-this
  before(): void { }

  // eslint-disable-next-line class-methods-use-this
  after(): void { }

  // eslint-disable-next-line class-methods-use-this
  hide(): void { }

  // eslint-disable-next-line class-methods-use-this
  show(): void { }

  remove(): void {
    const target = typeof this.el === 'string'
      ? document.querySelector(this.el)
      : this.el;

    if (target) {
      render(null, target);

      this.el = null;
    }
  }

  // eslint-disable-next-line class-methods-use-this
  getUser(): void { }

  on(eventName: OktaWidgetEventType, eventHandler: OktaWidgetEventHandler): void {
    this.events[eventName] = eventHandler;
  }

  // eslint-disable-next-line class-methods-use-this
  off(): void { }

  private buildRenderOptions(
    options: WidgetProps & Record<string, string> = {},
  ): RenderOptions {
    const widgetOptions = this.options;
    // @ts-ignore OKTA-508744
    const authParams: OktaAuthOptions = {
      ...widgetOptions.authParams,
      ...{
        responseType: options.responseType,
        scopes: options.scopes,
        state: options.state,
        nonce: options.nonce,
        idp: options.idp,
        idpScope: options.idpScope,
        display: options.display,
        prompt: options.prompt,
        maxAge: options.maxAge,
        loginHint: options.loginHint,
      },
    };

    const { el, clientId, redirectUri } = { ...widgetOptions, ...options };
    if (!el) {
      throw new Error('"el" is required');
    }
    if (!clientId) {
      throw new Error('"clientId" is required');
    }
    if (!redirectUri) {
      throw new Error('"redirectUri" is required');
    }
    return {
      el,
      clientId,
      redirectUri,
      authParams,
    };
  }
}
