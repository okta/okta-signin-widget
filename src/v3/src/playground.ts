/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

/* eslint-disable no-alert,no-console */

import { assertNoEnglishLeaks } from '../../../playground/LocaleUtils';
import {
  OktaSignInAPI,
  OktaSignInConstructor,
  RenderResult,
  RenderResultSuccessNonOIDCSession,
  WidgetOptions,
} from '../../types';
import { config } from '../widgetrc';
import OktaSignIn from './OktaSignIn'; // FIXME should be set by cdn bundle
import { getWorker } from './mocks/browser';

declare global {
  interface Window {
    // added by widget CDN bundle
    OktaSignIn: OktaSignInConstructor;

    // from <script src="/js/okta-plugin-a11y.js">
    OktaPluginA11y: { init: (widget: OktaSignInAPI) => void };

    // added in this file
    getWidgetInstance: () => OktaSignInAPI;
    createWidgetInstance: (options: WidgetOptions) => OktaSignInAPI;
    renderPlaygroundWidget: (options: WidgetOptions) => void;
    additionalOptions?: Partial<WidgetOptions>;
  }
}

function isSuccessNonOIDC(res: RenderResult): res is RenderResultSuccessNonOIDCSession {
  return (res as RenderResultSuccessNonOIDCSession).session !== undefined;
}

let siw: OktaSignInAPI;

function getWidgetInstance() {
  return siw;
}

function createWidgetInstance(options: WidgetOptions = {}) {
  if (siw) {
    siw.remove();
  }
  // @ts-expect-error mismatch between OktaSignIn and OktaSignInAPI
  siw = new OktaSignIn({ ...config, ...options });
  return siw;
}

const renderPlaygroundWidget = (options = {}) => {
  createWidgetInstance(options);

  siw.renderEl(
    { el: '#okta-login-container' },

    (res) => {
      // Password recovery flow
      if (res.status === 'FORGOT_PASSWORD_EMAIL_SENT') {
        alert('SUCCESS: Forgot password email sent');
        return;
      }

      // Unlock account flow
      if (res.status === 'UNLOCK_ACCOUNT_EMAIL_SENT') {
        alert('SUCCESS: Unlock account email sent');
        return;
      }

      // User has completed authentication (res.status === 'SUCCESS')

      if (isSuccessNonOIDC(res)) {
        // 1. Widget is not configured for OIDC, and returns a sessionToken that
        //    needs to be exchanged for an okta session
        console.log(res.user);
        res.session
          ?.setCookieAndRedirect(`${config.baseUrl}/app/UserHome`);
      } else if (Array.isArray(res)) {
        // 2. Widget is configured for OIDC, and returns tokens. This can be an
        //    array of tokens or a single token, depending on the initial
        //    configuration.
        console.log(JSON.stringify(res));
        alert('SUCCESS: OIDC with multiple responseTypes. Check console.');
      } else {
        console.log(JSON.stringify(res));
        alert('SUCCESS: OIDC with single responseType. Check Console');
      }
    },

    (err) => {
      console.error('global error handler: ', err);
    },
  ).catch(() => { /* we are using global error handler */ });

  siw.on('ready', () => {
    // handle `ready` event.
    // use `console.log` in particular so that those logs can be retrieved
    // in testcafe for assertion
    console.log('===== playground widget ready event received =====');
  });

  siw.on('afterRender', (context) => {
    // handle `afterRender` event.
    // use `console.log` in particular so that those logs can be retrieved
    // in testcafe for assertion
    console.log('===== playground widget afterRender event received =====');
    console.log(JSON.stringify(context));

    // assert english leaks if locale set to ok-PL
    if (config.language === 'ok-PL') {
      // Use innerText to avoid including hidden elements
      const viewText = document.getElementById('okta-sign-in')
        ?.innerText
        ?.split('\n').join(' ');

      const noTranslateElems = document.getElementsByClassName('no-translate');
      const noTranslationContent = [];
      if (noTranslateElems.length) {
        // eslint-disable-next-line no-plusplus
        for (let i = 0; i < noTranslateElems.length; i++) {
          // build array of noTranslationContent
          noTranslationContent.push(noTranslateElems[i].textContent);
        }
      }
      assertNoEnglishLeaks(null, viewText, noTranslationContent);
    }
  });

  siw.on('afterError', (context, error) => {
    // handle `afterError` event. use `console.log` in particular so that those
    // logs can be retrieved in testcafe for assertion
    console.log('===== playground widget afterError event received =====');
    console.log(JSON.stringify(context));
    console.log(JSON.stringify(error));
  });
};

window.getWidgetInstance = getWidgetInstance;
window.createWidgetInstance = createWidgetInstance;
window.renderPlaygroundWidget = renderPlaygroundWidget;

let render = true;

if (typeof URL !== 'undefined') {
  const { searchParams } = new URL(window.location.href);
  if (searchParams.get('render') === '0' || searchParams.get('render') === 'false') {
    render = false;
  }
} else {
  throw new Error('window.URL undefined');
}

// set up msw
getWorker()
  .then((worker) => worker?.start())
  .then(() => {
    if (render) {
      renderPlaygroundWidget(window.additionalOptions ?? {});
    }
  })
