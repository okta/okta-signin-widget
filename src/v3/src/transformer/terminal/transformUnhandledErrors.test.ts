/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { AuthApiError, OAuthError } from '@okta/okta-auth-js';
import { InfoboxElement, WidgetProps } from 'src/types';

import { transformUnhandledErrors } from './transformUnhandledErrors';

describe('Unhandled Error Transformer Tests', () => {
  describe('When AuthApiError is returned', () => {
    let apiError: AuthApiError;
    let widgetProps: WidgetProps;

    beforeEach(() => {
      widgetProps = {};
      apiError = {
        name: 'AuthApiError',
        message: '',
        errorSummary: '',
        errorCode: '',
      };
    });

    it('should add Infobox with unexpected error message when error is not provided', () => {
      const formBag = transformUnhandledErrors(widgetProps);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      const el = formBag.uischema.elements[0] as InfoboxElement;
      expect(el.type).toBe('InfoBox');
      expect(el.options?.message).toBe('oform.error.unexpected');
      expect(el.options?.class).toBe('ERROR');
    });

    it('should add infobox with custom message from server', () => {
      const mockErrorMessage = 'Custom error message';
      apiError = {
        ...apiError,
        errorCode: 'some_error_key',
        errorSummary: mockErrorMessage,
      };
      const formBag = transformUnhandledErrors(widgetProps, apiError);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      const el = formBag.uischema.elements[0] as InfoboxElement;
      expect(el.type).toBe('InfoBox');
      expect(el.options?.message).toBe(mockErrorMessage);
      expect(el.options?.class).toBe('ERROR');
    });
  });

  describe('When OAuthError is returned', () => {
    let apiError: OAuthError;
    let widgetProps: WidgetProps;

    beforeEach(() => {
      widgetProps = {};
      apiError = {
        name: 'OAuthError',
        message: '',
        errorCode: '',
        errorSummary: '',
        error: '',
        error_description: '',
      };
    });

    it('should add Infobox with unexpected error message when error is not provided', () => {
      const formBag = transformUnhandledErrors(widgetProps);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      const el = formBag.uischema.elements[0] as InfoboxElement;
      expect(el.type).toBe('InfoBox');
      expect(el.options?.message).toBe('oform.error.unexpected');
      expect(el.options?.class).toBe('ERROR');
    });

    it('should add info box when response is invalid recovery token error', () => {
      const mockErrorMessage = 'The recovery token is invalid';
      apiError = {
        ...apiError,
        error: 'invalid_request',
        error_description: mockErrorMessage,
      };
      const formBag = transformUnhandledErrors(widgetProps, apiError);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      expect(formBag.uischema.elements[0].type).toBe('InfoBox');
      expect((formBag.uischema.elements[0] as InfoboxElement).options?.message).toBe('oie.invalid.recovery.token');
      expect((
        formBag.uischema.elements[0] as InfoboxElement
      ).options?.class).toBe('ERROR');
    });

    it('should add info box when oie is not enabled error', () => {
      const mockErrorMessage = 'A mocked error message';
      apiError = {
        ...apiError,
        error: 'access_denied',
        error_description: mockErrorMessage,
      };
      const formBag = transformUnhandledErrors(widgetProps, apiError);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      const el = formBag.uischema.elements[0] as InfoboxElement;
      expect(el.type).toBe('InfoBox');
      expect(el.options?.message).toBe('oie.feature.disabled');
      expect(el.options?.class).toBe('ERROR');
    });

    it('should add info box when oie configuration error', () => {
      apiError = {
        ...apiError,
        error: 'unauthorized_client',
        error_description: 'The client is not authorized to use the provided grant type. Configured grant types: [refresh_token, implicit, authorization_code].',
      };
      const formBag = transformUnhandledErrors(widgetProps, apiError);

      expect(formBag).toMatchSnapshot();
      expect(formBag.uischema.elements.length).toBe(1);
      const el = formBag.uischema.elements[0] as InfoboxElement;
      expect(el.type).toBe('InfoBox');
      expect(el.options?.message).toBe('oie.configuration.error');
      expect(el.options?.class).toBe('ERROR');
    });
  });
});
