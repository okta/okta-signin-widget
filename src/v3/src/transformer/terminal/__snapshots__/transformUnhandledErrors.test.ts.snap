// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unhandled Error Transformer Tests When AuthApiError is returned should add Infobox with unexpected error message when error is not provided 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "oform.error.unexpected",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests When AuthApiError is returned should add infobox with custom message from server 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "Custom error message",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests When OAuthError is returned should add Infobox with unexpected error message when error is not provided 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "oform.error.unexpected",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests When OAuthError is returned should add info box when oie configuration error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "oie.configuration.error",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests When OAuthError is returned should add info box when oie is not enabled error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "oie.feature.disabled",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Unhandled Error Transformer Tests When OAuthError is returned should add info box when response is invalid recovery token error 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToValidate": Array [],
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "class": "ERROR",
          "dataSe": "callout",
          "message": "oie.invalid.recovery.token",
        },
        "type": "InfoBox",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
